src/gamemode.c

```diff
@@ -0,0 +1,341 @@
+  gboolean ok;
+
+  ok = xdp_impl_permission_store_call_lookup_sync (get_permission_store (),
+                                                   PERMISSION_TABLE,
+                                                   PERMISSION_ID,
+                                                   &perms,
+                                                   &data,
+                                                   NULL,
+                                                   &err);
+
+  if (!ok)
+    {
+      g_dbus_error_strip_remote_error (err);
+      g_debug ("No gamemode permissions found: %s", err->message);
+      g_clear_error (&err);
+    }
+  else if (perms != NULL && g_variant_lookup (perms, app_id, "^a&s", &stored))
+    {
+      g_autofree char *as_str = NULL;
+      gboolean allowed;
+
+      as_str = g_strjoinv (" ", (char **)stored);
+      g_debug ("Notification gamemode for %s: %s", app_id, as_str);
+
+      allowed = !g_strv_contains (stored, "no");
+
+      if (!allowed)
+	g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+		     "GameMode is not allowed for %s", app_id);
+
+      return allowed;
+    }
+
+  g_debug ("No gamemode permissions stored for %s: allowing", app_id);
+
+  ok = xdp_impl_permission_store_call_set_permission_sync (get_permission_store(),
+							   PERMISSION_TABLE,
+							   TRUE,
+							   PERMISSION_ID,
+							   app_id,
+							   (const char*[]) {"yes", NULL},
+							   NULL,
+							   &err);
+  if (!ok)
+    {
+      g_dbus_error_strip_remote_error (err);
+      g_warning ("Error updating permission store: %s", err->message);
+    }
+
+  return TRUE;
+}
+
+/* generic dbus call handling */
+
+typedef struct CallData_ {
+  GDBusMethodInvocation *inv;
+  XdpAppInfo *app_info;
+
+  char *method;
+  gint  pid;
+
+} CallData;
+
+static CallData *
+call_data_new (GDBusMethodInvocation *inv,
+	       XdpAppInfo            *app_info,
+	       const char            *method)
+{
+  CallData *call;
+
+  call = g_slice_new0 (CallData);
+
+  call->inv = g_object_ref (inv);
+  call->app_info = xdp_app_info_ref (app_info);
+  call->method = g_strdup (method);
+
+  return call;
+}
+
+static void
+call_data_free (gpointer data)
+{
+  CallData *call = data;
+  if (call == NULL)
+    return;
+
+  g_object_unref (call->inv);
+  xdp_app_info_unref (call->app_info);
+  g_free (call->method);
+
+  g_slice_free (CallData, call);
+}
+
+static void
+handle_call_thread (GTask        *task,
+                    gpointer      source_object,
+                    gpointer      task_data,
+                    GCancellable *cancellable)
+{
+  g_autoptr(GError) err = NULL;
+  g_autoptr(GVariant) res = NULL;
+  const char *app_id;
+  CallData *call;
+  pid_t mapped;
+  gboolean ok;
+  gint r;
+
+  call = (CallData *) task_data;
+  app_id = xdp_app_info_get_id (call->app_info);
+
+  if (!gamemode_is_allowed_for_app (app_id, &err))
+    {
+      g_dbus_method_invocation_take_error (call->inv, g_steal_pointer (&err));
```

Comment:
I think we generally use
```
g_dbus_method_invocation_return_gerror (invocation, error);
```
for this. And I would slightly prefer "error" as variable name.